Add sprint 1 iteration 5

a-tho opened the pull request.

a-tho committed:
Add sprint 1 increment 1

a-tho committed:
Minor fix

a-tho committed:
Do not treat missing content-type as invalid req

a-tho committed:
Refactor to allow 404 when no metric name provided

a-tho committed:
Use errors.New instead of fmt.Errorf

a-tho committed:
Add sprint 1 iteration 2

Added a client (an agent) that collects metrics and transmits them to
the server.

a-tho committed:
Fix typos

a-tho committed:
Minor fixes

a-tho committed:
Add tests for server package

a-tho committed:
Add tests for storage package

a-tho committed:
Fix typo

a-tho committed:
Add sprint 1 iteration 3

Use chi and resty instead of net/http for server and client
implementations.

a-tho committed:
Remove redundant files

a-tho committed:
Update sprint 1 iteration 3

Refactored the server and the client to enable curly-brackets-routing
for /update handler.
curly brackets for that).

a-tho committed:
Fix tests for refactored server

a-tho committed:
Fix statictest error

Response body is required to be closed where the life of response
object ends, not where response body is used the last time.

a-tho committed:
Add metrics retrieval handlers

a-tho committed:
Fix typo in tests

a-tho committed:
Add flags

a-tho committed:
go mod tidy

a-tho committed:
Move parseFlags to main.go for agent

a-tho committed:
Fix bugs in agent

 - Add PollCount collection
 - Add error handling in Observe()
 - Add missing protocol for connections to server

a-tho committed:
Add parsing config vars from env vars

a-tho committed:
Fix flawed type casts for time itervals

rktkov commented on internal/telemetry/telemetry.go:
```diff
@@ -0,0 +1,111 @@
+// Package telemetry implements a trivial way to collect and transmit metrics.
+package telemetry
+
+import (
```
Возьмите в привычку сортировать импортируемые пакеты следующим образом:
1. пакеты стандартной библиотеки
2. внешние установленные пакеты
3. внутренние пакеты проекта

Для того, чтобы разные автоматические инструменты их не пересортировывали, разделяйте блоки пустой строкой.

см. комментарий про порядок импорта
https://stackoverflow.com/questions/24790175/when-is-the-init-function-run

rktkov commented on internal/telemetry/telemetry.go:
```diff
@@ -0,0 +1,111 @@
+	polled []monitor.MetricInstance
+}
+
+func New(srvAddr string, pollInterval, reportStep int) *Observer {
```
Лучше назвать метод NewObserver

rktkov commented on internal/telemetry/telemetry.go:
```diff
@@ -0,0 +1,111 @@
+			// Report to the server
+			for _, instance := range o.polled {
+				// Gauge metrics
+				for key, value := range instance.Gauges {
```
Действия сбора метрики и отправки стоит разделить. У нас ведь разные интервалы для каждого из них настраиваются.

rktkov commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+// New returns an initialized storage.
+func New[T monitor.Gauge | monitor.Counter]() *MemStorage[T] {
```
Стоит создать отдельную функцию инициализатора для хранилища метрик и хранилища коунтера. А вообще использование дженериков на данном этапе их реализации в ГО является тернистым путем. Я бы рекомендовал переписать этот код без использования их.

rktkov commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,66 @@
+	return string(out)
+}
+
+func (s MemStorage[T]) HTML() (*bytes.Buffer, error) {
```
Этот функционал выглядит лишним для хранилища, этим должен заниматься клиент хранилища, а не оно само.

rktkov commented on internal/server/server.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+// New creates a new multiplexer with configured handlers
+func New(
```
Стоит переименовать в NewServer

rktkov commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,105 @@
+
+	w.Write([]byte(pageHead))
+	w.Write([]byte(gaugeHeader))
+	w.Write(gaugeBuf.Bytes())
```
Генерацию HTML вот куда-то сюда лучше перенести.

a-tho committed:
Fix imports

a-tho committed:
Try to fix precision for GetValHandler

a-tho committed:
Fix tests

a-tho committed:
Revert precision changes for GetValHandler

a-tho committed:
Try to fix precision for GetValHandler

a-tho committed:
Update factory functions for server and telemetry

a-tho committed:
Update server and storage packages

 - Removed generics from metrics storage implementation
 - Moved HTML outputting of metrics storage data to the client
   (handlers.go)

a-tho committed:
Split polling and reporting logic

rktkov commented on internal/telemetry/telemetry.go:
```diff
@@ -0,0 +1,121 @@
+
+func (o *Observer) poll(pollCount int) {
+	var (
+		t        = pollCount % o.reportStep
```
Стоит выбирать более говорящие имена для переменных.

a-tho committed:
Rename index variable in telemetry.poll()
